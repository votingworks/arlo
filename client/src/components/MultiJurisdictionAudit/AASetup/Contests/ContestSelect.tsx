/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { useState } from 'react'
import { useParams } from 'react-router-dom'
import { Formik, FormikProps, getIn } from 'formik'
import { Checkbox, Spinner } from '@blueprintjs/core'
import { Column, Cell } from 'react-table'
import FormWrapper from '../../../Atoms/Form/FormWrapper'
import FormButtonBar from '../../../Atoms/Form/FormButtonBar'
import FormButton from '../../../Atoms/Form/FormButton'
import { ISidebarMenuItem } from '../../../Atoms/Sidebar'
import useStandardizedContests, {
  IStandardizedContest,
  IStandardizedContestOption,
} from '../../useStandardizedContests'
import useJurisdictions from '../../useJurisdictions'
import { IAuditSettings } from '../../../../types'
import { Table, FilterInput } from '../../../Atoms/Table'

interface IProps {
  isTargeted: boolean
  nextStage: ISidebarMenuItem
  prevStage: ISidebarMenuItem
  locked: boolean
  auditType: IAuditSettings['auditType']
}

const ContestSelect: React.FC<IProps> = ({
  isTargeted,
  nextStage,
  prevStage,
  //   locked,
}) => {
  const { electionId } = useParams<{ electionId: string }>()
  const [standardizedContests, updateContests] = useStandardizedContests(
    electionId,
    isTargeted
  )
  const [filter, setFilter] = useState('')
  const jurisdictions = useJurisdictions(electionId)

  if (!standardizedContests || !jurisdictions) return null // Still loading

  const submit = async ({
    contests,
  }: {
    contests: IStandardizedContestOption[]
  }) => {
    const response = await updateContests(contests)
    // TEST TODO
    /* istanbul ignore next */
    if (!response) return
    /* istanbul ignore else */
    if (nextStage.activate) nextStage.activate()
    else throw new Error('Wrong menuItems passed in: activate() is missing')
  }

  // TODO filter by jurisdiction names as well
  const filteredContests = standardizedContests.filter(({ name }) =>
    name.toLowerCase().includes(filter.toLowerCase())
  )

  const columns = (
    values: { contests: IStandardizedContest[] },
    setFieldValue: FormikProps<{
      contests: IStandardizedContest[]
    }>['setFieldValue']
  ): Column<IStandardizedContest>[] => [
    {
      id: 'select',
      Header: 'Select',
      accessor: row => row,
      // eslint-disable-next-line react/display-name
      Cell: ({
        row: { original: contest, index },
      }: Cell<IStandardizedContest>) => (
        <Checkbox
          inline
          checked={getIn(values, `contests[${index}].checked`)}
          onChange={({
            currentTarget: { checked },
          }: React.ChangeEvent<HTMLInputElement>) =>
            setFieldValue(`contests[${index}]`, { ...contest, checked })
          }
          disabled={standardizedContests[index].isTargeted !== isTargeted}
        />
      ),
    },
    {
      Header: 'Contest Name',
      accessor: 'name',
    },
    {
      Header: 'Jurisdiction(s)',
      accessor: row =>
        row.jurisdictionIds.length === jurisdictions.length
          ? 'All'
          : row.jurisdictionIds
              .map(id => jurisdictions.find(j => j.id === id)!.name)
              .join(' - '),
    },
  ]

  return (
    <Formik
      initialValues={{
        contests: standardizedContests,
      }}
      enableReinitialize
      onSubmit={submit}
    >
      {({
        values,
        handleSubmit,
        setFieldValue,
      }: FormikProps<{ contests: IStandardizedContest[] }>) => (
        <form>
          <FormWrapper
            title={isTargeted ? 'Target Contests' : 'Opportunistic Contests'}
          >
            <p>
              Choose which contests to{' '}
              {isTargeted ? 'target for audit' : 'audit opportunistically'} by
              checking the checkboxes below. To filter the contest list, use the
              search box at the top.
            </p>
            <FilterInput
              placeholder="Filter by contest or jurisdiction name..."
              value={filter}
              onChange={value => setFilter(value)}
            />
            <Table
              data={filteredContests}
              columns={columns(values, setFieldValue)}
            />
          </FormWrapper>
          {nextStage.state === 'processing' ? (
            <Spinner />
          ) : (
            <FormButtonBar>
              <FormButton onClick={prevStage.activate}>Back</FormButton>
              <FormButton
                type="submit"
                intent="primary"
                disabled={nextStage.state === 'locked'}
                onClick={handleSubmit}
              >
                Save &amp; Next
              </FormButton>
            </FormButtonBar>
          )}
        </form>
      )}
    </Formik>
  )
}

export default ContestSelect
