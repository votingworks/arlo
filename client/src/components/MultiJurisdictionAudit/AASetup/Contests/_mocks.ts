import { IContest } from '../../../../types'

export const contestMocks: {
  [key: string]: { contests: IContest[] }
} = {
  emptyTargeted: {
    contests: [
      {
        id: '',
        name: '',
        isTargeted: true,
        totalBallotsCast: '',
        numWinners: '1',
        votesAllowed: '1',
        jurisdictionIds: [],
        choices: [
          {
            id: '',
            name: '',
            numVotes: '',
          },
          {
            id: '',
            name: '',
            numVotes: '',
          },
        ],
      },
    ],
  },
  emptyOpportunistic: {
    contests: [
      {
        id: '',
        name: '',
        isTargeted: false,
        totalBallotsCast: '',
        numWinners: '1',
        votesAllowed: '1',
        jurisdictionIds: [],
        choices: [
          {
            id: '',
            name: '',
            numVotes: '',
          },
          {
            id: '',
            name: '',
            numVotes: '',
          },
        ],
      },
    ],
  },
  filledTargeted: {
    contests: [
      {
        id: 'contest-id',
        name: 'Contest Name',
        isTargeted: true,
        totalBallotsCast: '30',
        numWinners: '1',
        votesAllowed: '1',
        jurisdictionIds: [],
        choices: [
          {
            id: 'choice-id-1',
            name: 'Choice One',
            numVotes: '10',
          },
          {
            id: 'choice-id-2',
            name: 'Choice Two',
            numVotes: '20',
          },
        ],
      },
    ],
  },
  filledOpportunistic: {
    contests: [
      {
        id: 'contest-id',
        name: 'Contest Name',
        isTargeted: false,
        totalBallotsCast: '30',
        numWinners: '1',
        votesAllowed: '1',
        jurisdictionIds: [],
        choices: [
          {
            id: 'choice-id-3',
            name: 'Choice Three',
            numVotes: '10',
          },
          {
            id: 'choice-id-4',
            name: 'Choice Four',
            numVotes: '20',
          },
        ],
      },
    ],
  },
  filledTargetedWithJurisdictionId: {
    contests: [
      {
        id: 'contest-id',
        name: 'Contest Name',
        isTargeted: true,
        totalBallotsCast: '30',
        numWinners: '1',
        votesAllowed: '1',
        jurisdictionIds: ['jurisdiction-id-1', 'jurisdiction-id-2'],
        choices: [
          {
            id: 'choice-id-1',
            name: 'Choice One',
            numVotes: '10',
          },
          {
            id: 'choice-id-2',
            name: 'Choice Two',
            numVotes: '20',
          },
        ],
      },
    ],
  },
  filledOpportunisticWithJurisdictionId: {
    contests: [
      {
        id: 'contest-id',
        name: 'Contest Name',
        isTargeted: false,
        totalBallotsCast: '30',
        numWinners: '1',
        votesAllowed: '1',
        jurisdictionIds: ['jurisdiction-id-1', 'jurisdiction-id-2'],
        choices: [
          {
            id: 'choice-id-3',
            name: 'Choice Three',
            numVotes: '10',
          },
          {
            id: 'choice-id-4',
            name: 'Choice Four',
            numVotes: '20',
          },
        ],
      },
    ],
  },
  filledTargetedAndOpportunistic: {
    contests: [
      {
        id: 'contest-id',
        name: 'Contest 1',
        isTargeted: true,
        totalBallotsCast: '30',
        numWinners: '1',
        votesAllowed: '1',
        jurisdictionIds: ['jurisdiction-id-1', 'jurisdiction-id-2'],
        choices: [
          {
            id: 'choice-id-1',
            name: 'Choice One',
            numVotes: '10',
          },
          {
            id: 'choice-id-2',
            name: 'Choice Two',
            numVotes: '20',
          },
        ],
      },
      {
        id: 'contest-id',
        name: 'Contest 2',
        isTargeted: false,
        totalBallotsCast: '30',
        numWinners: '1',
        votesAllowed: '1',
        jurisdictionIds: ['jurisdiction-id-1', 'jurisdiction-id-2'],
        choices: [
          {
            id: 'choice-id-3',
            name: 'Choice Three',
            numVotes: '10',
          },
          {
            id: 'choice-id-4',
            name: 'Choice Four',
            numVotes: '20',
          },
        ],
      },
    ],
  },
}

export const contestsInputMocks = {
  inputs: [
    { key: 'Contest Name', value: 'Contest Name' },
    { key: 'Name of Candidate/Choice 1', value: 'Choice One' },
    { key: 'Name of Candidate/Choice 2', value: 'Choice Two' },
    { key: 'Votes for Candidate/Choice 1', value: '10' },
    { key: 'Votes for Candidate/Choice 2', value: '20' },
    { key: 'Total Ballots for Contest', value: '30' },
  ],
  errorInputs: [
    { key: 'Contest Name', value: '', error: 'Required' },
    {
      key: 'Total Ballots for Contest',
      value: '',
      error:
        'Must be greater than or equal to the sum of votes for each candidate/choice',
    },
    {
      key: 'Total Ballots for Contest',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Total Ballots for Contest',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Total Ballots for Contest',
      value: '0.5',
      error: 'Must be an integer',
    },
    { key: 'Name of Candidate/Choice 1', value: '', error: 'Required' },
    { key: 'Name of Candidate/Choice 2', value: '', error: 'Required' },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '',
      error: 'Required',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '0.5',
      error: 'Must be an integer',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '',
      error: 'Required',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '0.5',
      error: 'Must be an integer',
    },
  ],
}

export default contestsInputMocks
