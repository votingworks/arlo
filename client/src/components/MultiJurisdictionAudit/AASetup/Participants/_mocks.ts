import { readFileSync } from 'fs'
import { join } from 'path'

const jurisdictionFile = new File(
  [
    readFileSync(
      join(
        __dirname,
        '../../../../../public/sample_jurisdiction_filesheet.csv'
      ),
      'utf8'
    ),
  ],
  'jurisdictions.csv',
  { type: 'text/csv' }
)

export default jurisdictionFile
