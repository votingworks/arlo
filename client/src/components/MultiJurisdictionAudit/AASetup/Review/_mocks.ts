import { IAuditSettings } from '../../../../types'

export const settingsMock: {
  [key in 'empty' | 'full' | 'offline' | 'batch']: IAuditSettings
} = {
  empty: {
    state: null,
    electionName: null,
    online: null,
    randomSeed: null,
    riskLimit: null,
    auditType: 'BALLOT_POLLING',
    auditName: 'Test Audit',
  },
  full: {
    state: 'AL',
    electionName: 'Election Name',
    online: true,
    randomSeed: '12345',
    riskLimit: 10,
    auditType: 'BALLOT_POLLING',
    auditName: 'Test Audit',
  },
  offline: {
    state: 'AL',
    electionName: 'Election Name',
    online: false,
    randomSeed: '12345',
    riskLimit: 10,
    auditType: 'BALLOT_POLLING',
    auditName: 'Test Audit',
  },
  batch: {
    state: 'AL',
    electionName: 'Election Name',
    online: false,
    randomSeed: '12345',
    riskLimit: 10,
    auditType: 'BATCH_COMPARISON',
    auditName: 'Test Audit',
  },
}

export const sampleSizeMock = {
  sampleSizes: {
    'contest-id': [
      { prob: 0.54, size: 46, key: 'asn' },
      { prob: 0.7, size: 67, key: '0.7' },
      { prob: 0.5, size: 88, key: '0.5' },
      { prob: 0.9, size: 125, key: '0.9' },
    ],
  },
}

export default settingsMock
