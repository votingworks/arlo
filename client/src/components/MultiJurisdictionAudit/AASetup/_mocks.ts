import { ElementType } from '../../../types'
import { setupStages, stageTitles } from './index'
import { ISidebarMenuItem } from '../../Atoms/Sidebar'

const relativeStages = (
  stage: ElementType<typeof setupStages>,
  state: ISidebarMenuItem['state'] = 'live'
): {
  prevStage: ISidebarMenuItem
  nextStage: ISidebarMenuItem
  menuItems: ISidebarMenuItem[]
} => {
  const prevId = setupStages[setupStages.indexOf(stage) - 1]
  const prevStage: ISidebarMenuItem = {
    id: prevId,
    title: stageTitles[prevId],
    active: false,
    activate: jest.fn(),
    state,
  }
  const nextId = setupStages[setupStages.indexOf(stage) + 1]
  const nextStage: ISidebarMenuItem = {
    id: nextId,
    title: stageTitles[nextId],
    active: false,
    activate: jest.fn(),
    state,
  }
  const menuItems = setupStages.map((s: ElementType<typeof setupStages>) => ({
    id: s,
    title: stageTitles[s],
    active: s === stage,
    activate: jest.fn(),
    state,
  }))
  return { prevStage, nextStage, menuItems }
}

export default relativeStages
