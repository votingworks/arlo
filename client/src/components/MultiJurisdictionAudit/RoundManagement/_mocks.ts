export interface INullResultValues {
  [contestId: string]: {
    [choiceId: string]: null | string | number
  }
}

export const resultsMocks: {
  [key in 'emptyInitial' | 'complete']: INullResultValues
} = {
  emptyInitial: {
    'contest-id-1': {
      'choice-id-1': null,
      'choice-id-2': null,
    },
    'contest-id-2': {
      'choice-id-3': null,
      'choice-id-4': null,
    },
  },
  complete: {
    'contest-id-1': {
      'choice-id-1': 1,
      'choice-id-2': 2,
    },
    'contest-id-2': {
      'choice-id-3': 1,
      'choice-id-4': 2,
    },
  },
}

export default resultsMocks
