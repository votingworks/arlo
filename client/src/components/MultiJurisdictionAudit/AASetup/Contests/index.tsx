/* eslint-disable jsx-a11y/label-has-associated-control */
import React from 'react'
import { ISidebarMenuItem } from '../../../Atoms/Sidebar'
import { IAuditSettings } from '../../../../types'
import ContestForm from './ContestForm'
import ContestSelect from './ContestSelect'

interface IProps {
  isTargeted: boolean
  nextStage: ISidebarMenuItem
  prevStage: ISidebarMenuItem
  locked: boolean
  auditType: IAuditSettings['auditType']
}

const Contests: React.FC<IProps> = (props: IProps) => {
  return props.auditType === 'BALLOT_COMPARISON' ? (
    <ContestSelect {...props} />
  ) : (
    <ContestForm {...props} />
  )
}

export default Contests
