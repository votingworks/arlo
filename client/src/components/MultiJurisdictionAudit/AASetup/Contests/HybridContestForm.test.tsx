import React from 'react'
import { Route } from 'react-router-dom'
import { fireEvent, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
  renderWithRouter,
  withMockFetch,
  regexpEscape,
} from '../../../testUtilities'
import relativeStages from '../_mocks'
import Contests from '.'
import { aaApiCalls } from '../../_mocks'
import { IContest, INewContest } from '../../useContestsBallotComparison'

const hybridContestsInputMocks = {
  inputs: [
    { key: 'Name of Candidate/Choice 1', value: 'Choice One' },
    { key: 'Name of Candidate/Choice 2', value: 'Choice Two' },
    { key: 'Votes for Candidate/Choice 1', value: '10' },
    { key: 'Votes for Candidate/Choice 2', value: '20' },
  ],
  errorInputs: [
    { key: 'Name of Candidate/Choice 1', value: '', error: 'Required' },
    { key: 'Name of Candidate/Choice 2', value: '', error: 'Required' },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '',
      error: 'Required',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '0.5',
      error: 'Must be an integer',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '',
      error: 'Required',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '0.5',
      error: 'Must be an integer',
    },
  ],
}

const { nextStage, prevStage } = relativeStages('target-contests')

const render = (isTargeted: boolean = true) =>
  renderWithRouter(
    <Route path="/election/:electionId/setup">
      <Contests
        auditType="HYBRID"
        locked={false}
        isTargeted={isTargeted}
        nextStage={nextStage}
        prevStage={prevStage}
      />
    </Route>,
    { route: '/election/1/setup' }
  )

const apiCalls = {
  getStandardizedContests: {
    url: '/api/election/1/standardized-contests',
    response: [
      {
        name: 'Contest 1.\'"', // Make sure dots and quotes in the name work
        jurisdictionIds: ['jurisdiction-id-1', 'jurisdiction-id-2'],
      },
      {
        name: 'Contest 2',
        jurisdictionIds: ['jurisdiction-id-1'],
      },
      { name: 'Contest 3', jurisdictionIds: ['jurisdiction-id-2'] },
    ],
  },
  getContests: (contests: IContest[]) => ({
    url: '/api/election/1/contest',
    response: { contests },
  }),
  putContests: (contests: INewContest[]) => ({
    url: '/api/election/1/contest',
    options: {
      method: 'PUT',
      body: JSON.stringify(contests),
      headers: { 'Content-Type': 'application/json' },
    },
    response: { status: 'ok' },
  }),
  submitContests: (mockContest: object) => ({
    url: '/api/election/1/contest',
    options: {
      method: 'PUT',
      body: JSON.stringify(mockContest),
      headers: { 'Content-Type': 'application/json' },
    },
    response: { status: 'ok' },
  }),
}

jest.mock('uuidv4', () => {
  let id = 0
  return () => {
    id += 1
    return id.toString()
  }
})

function typeInto(input: Element, value: string): void {
  // TODO: do more, like `focusIn`, `focusOut`, `input`, etc?
  fireEvent.focus(input)
  fireEvent.change(input, { target: { value } })
  fireEvent.blur(input)
}

describe('Audit Setup > Contests (Hybrid)', () => {
  let id = 0
  const getID = () => {
    id += 1
    return id.toString()
  }

  // created function to generate new IDs
  const newContest = () => {
    return [
      {
        id: getID(),
        name: 'Contest 1.\'"',
        isTargeted: true,
        numWinners: 1,
        votesAllowed: 1,
        jurisdictionIds: ['jurisdiction-id-1', 'jurisdiction-id-2'],
        choices: [
          {
            id: getID(),
            name: 'Choice One',
            numVotes: 10,
          },
          {
            id: getID(),
            name: 'Choice Two',
            numVotes: 20,
          },
        ],
      },
    ]
  }

  it('Audit Setup > Contests', async () => {
    const expectedCalls = [
      apiCalls.getContests([]),
      aaApiCalls.getJurisdictions,
      apiCalls.getStandardizedContests,
    ]
    await withMockFetch(expectedCalls, async () => {
      const { container } = render()
      await screen.findByRole('heading', { name: 'Target Contests' })
      expect(container).toMatchSnapshot()
    })
  })

  it('is able to submit form successfully', async () => {
    const expectedCalls = [
      apiCalls.getContests([]),
      aaApiCalls.getJurisdictions,
      apiCalls.getStandardizedContests,
      apiCalls.submitContests(newContest()),
    ]
    await withMockFetch(expectedCalls, async () => {
      const { getByLabelText } = render()
      await screen.findByRole('heading', { name: 'Target Contests' })
      userEvent.selectOptions(
        screen.getByLabelText(/Contest Name/),
        'Contest 1.\'"'
      )
      hybridContestsInputMocks.inputs.forEach(inputData => {
        const input = getByLabelText(new RegExp(regexpEscape(inputData.key)), {
          selector: 'input',
        }) as HTMLInputElement
        typeInto(input, inputData.value)
        expect(input.value).toBe(inputData.value)
      })
      expect(screen.queryByText('Total Ballots Cast')).not.toBeInTheDocument()
      userEvent.click(screen.getByRole('button', { name: 'Save & Next' }))
      await waitFor(() => expect(nextStage.activate).toHaveBeenCalled())
    })
  })

  it('Check Jurisdiction selection is hidden for hybrid', async () => {
    const expectedCalls = [
      apiCalls.getContests([]),
      aaApiCalls.getJurisdictions,
      apiCalls.getStandardizedContests,
      apiCalls.submitContests(newContest()),
    ]
    await withMockFetch(expectedCalls, async () => {
      const { getByLabelText } = render()
      await screen.findByRole('heading', { name: 'Target Contests' })
      userEvent.selectOptions(
        screen.getByLabelText(/Contest Name/),
        'Contest 1.\'"'
      )
      hybridContestsInputMocks.inputs.forEach(inputData => {
        const input = getByLabelText(new RegExp(regexpEscape(inputData.key)), {
          selector: 'input',
        }) as HTMLInputElement
        typeInto(input, inputData.value)
        expect(input.value).toBe(inputData.value)
      })
      expect(screen.queryByText('Contest Universe')).not.toBeInTheDocument()
      userEvent.click(screen.getByRole('button', { name: 'Save & Next' }))
      await waitFor(() => expect(nextStage.activate).toHaveBeenCalled())
    })
  })

  it('it should not skip to next stage when targeted contest form is clean and not touched', async () => {
    const expectedCalls = [
      apiCalls.getContests([]),
      aaApiCalls.getJurisdictions,
      apiCalls.getStandardizedContests,
    ]
    await withMockFetch(expectedCalls, async () => {
      render()
      await screen.findByRole('heading', { name: 'Target Contests' })
      userEvent.click(screen.getByRole('button', { name: 'Save & Next' }))
      await waitFor(() => {
        expect(screen.queryAllByText('Required').length).toBe(5)
      })
    })
  })

  it('it should skip to next stage when opportunistic contest form is clean and not touched', async () => {
    const expectedCalls = [
      apiCalls.getContests([]),
      aaApiCalls.getJurisdictions,
      apiCalls.getStandardizedContests,
    ]
    await withMockFetch(expectedCalls, async () => {
      render(false)

      await screen.findByRole('heading', { name: 'Opportunistic Contests' })
      userEvent.click(screen.getByRole('button', { name: 'Save & Next' }))
      await waitFor(() => expect(nextStage.activate).toHaveBeenCalled())
    })
  })

  it('it should not skip to next stage when opportunistic contest form is touched', async () => {
    const expectedCalls = [
      apiCalls.getContests([]),
      aaApiCalls.getJurisdictions,
      apiCalls.getStandardizedContests,
    ]
    await withMockFetch(expectedCalls, async () => {
      const { getByLabelText } = render(false)

      await screen.findByRole('heading', { name: 'Opportunistic Contests' })

      typeInto(
        getByLabelText('Votes Allowed', {
          selector: 'input',
        }),
        '2'
      )

      userEvent.click(screen.getByRole('button', { name: 'Save & Next' }))

      await waitFor(() => {
        expect(screen.queryAllByText('Required').length).toBe(5)
      })
    })
  })
})
