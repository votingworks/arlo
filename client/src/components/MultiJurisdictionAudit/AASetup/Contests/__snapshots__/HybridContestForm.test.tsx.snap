// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Audit Setup > Contests (Hybrid) Audit Setup > Contests 1`] = `
<div>
  <form
    data-testid="form-one"
  >
    <div
      class="sc-EHOje bjGJmd"
    >
      <h2
        class="bp3-heading sc-ifAKCX cmZyyF"
      >
        Target Contests
      </h2>
      <div
        class="bp3-card bp3-elevation-2 sc-dxgOiQ jDdoJq"
      >
        <div
          class="sc-htoDjs jcQPAQ"
        >
          <h5
            class="bp3-heading"
          >
            Contest  Info
          </h5>
          <br />
          <div>
            <div
              class="sc-dnqmqq kGMqKN"
            >
              Select the name of the contest that will drive the audit.
            </div>
            <label
              for="contests[0].name"
            >
              Contest
               
              
               Name
              <br />
              <div
                class="bp3-html-select sc-ckVGcZ hwbwFM"
              >
                <select
                  field="[object Object]"
                  form="[object Object]"
                  id="contests[0].name"
                >
                  <option
                    value=""
                  />
                  <option
                    label="Contest 1.'\\""
                    value="Contest 1.'\\""
                  >
                    Contest 1.'"
                  </option>
                  <option
                    label="Contest 2"
                    value="Contest 2"
                  >
                    Contest 2
                  </option>
                  <option
                    label="Contest 3"
                    value="Contest 3"
                  >
                    Contest 3
                  </option>
                </select>
                <span
                  class="bp3-icon bp3-icon-double-caret-vertical"
                  icon="double-caret-vertical"
                >
                  <svg
                    data-icon="double-caret-vertical"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      double-caret-vertical
                    </desc>
                    <path
                      d="M5 7h6a1.003 1.003 0 00.71-1.71l-3-3C8.53 2.11 8.28 2 8 2s-.53.11-.71.29l-3 3A1.003 1.003 0 005 7zm6 2H5a1.003 1.003 0 00-.71 1.71l3 3c.18.18.43.29.71.29s.53-.11.71-.29l3-3A1.003 1.003 0 0011 9z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </div>
            </label>
          </div>
          <div
            class="sc-dnqmqq kGMqKN"
          >
            Enter the number of winners for the contest.
          </div>
          <label
            for="contests[0].numWinners"
          >
            Winners
            <div
              class="sc-gZMcBi ggdJUe"
            >
              <div
                class="bp3-input-group sc-gqjmRU kmAUGh"
              >
                <input
                  class="bp3-input"
                  id="contests[0].numWinners"
                  name="contests[0].numWinners"
                  style="padding-right: 10px;"
                  value="1"
                />
              </div>
            </div>
          </label>
          <div
            class="sc-dnqmqq kGMqKN"
          >
            Number of selections the voter can make in the contest.
          </div>
          <label
            for="contests[0].votesAllowed"
          >
            Votes Allowed
            <div
              class="sc-gZMcBi ggdJUe"
            >
              <div
                class="bp3-input-group sc-gqjmRU kmAUGh"
              >
                <input
                  class="bp3-input"
                  id="contests[0].votesAllowed"
                  name="contests[0].votesAllowed"
                  style="padding-right: 10px;"
                  value="1"
                />
              </div>
            </div>
          </label>
        </div>
        <div
          class="sc-htoDjs jcQPAQ"
        >
          <h5
            class="bp3-heading"
          >
            Candidates/Choices & Vote Totals
          </h5>
          <div
            class="sc-dnqmqq kGMqKN"
          >
            Enter the name of each candidate choice that appears on the ballot for this contest.
          </div>
          <div
            class="sc-cSHVUG cNFbaH"
          >
            <div
              class="sc-kAzzGY gvpNKy"
            >
              <label
                class="bp3-label sc-kgoBCf eKfxQS"
              >
                Name of Candidate/Choice 
                1
                <div
                  class="sc-chPdSV knRDOr sc-gZMcBi ggdJUe"
                >
                  <div
                    class="bp3-input-group sc-gqjmRU kmAUGh"
                  >
                    <input
                      class="bp3-input"
                      name="contests[0].choices[0].name"
                      style="padding-right: 10px;"
                      value=""
                    />
                  </div>
                </div>
              </label>
              <label
                class="bp3-label sc-kgoBCf eKfxQS"
              >
                Votes for Candidate/Choice 
                1
                <div
                  class="sc-chPdSV knRDOr sc-gZMcBi ggdJUe"
                >
                  <div
                    class="bp3-input-group sc-gqjmRU kmAUGh"
                  >
                    <input
                      class="bp3-input"
                      name="contests[0].choices[0].numVotes"
                      style="padding-right: 10px;"
                      value=""
                    />
                  </div>
                </div>
              </label>
            </div>
            <div
              class="sc-kAzzGY gvpNKy"
            >
              <label
                class="bp3-label sc-kgoBCf eKfxQS"
              >
                Name of Candidate/Choice 
                2
                <div
                  class="sc-chPdSV knRDOr sc-gZMcBi ggdJUe"
                >
                  <div
                    class="bp3-input-group sc-gqjmRU kmAUGh"
                  >
                    <input
                      class="bp3-input"
                      name="contests[0].choices[1].name"
                      style="padding-right: 10px;"
                      value=""
                    />
                  </div>
                </div>
              </label>
              <label
                class="bp3-label sc-kgoBCf eKfxQS"
              >
                Votes for Candidate/Choice 
                2
                <div
                  class="sc-chPdSV knRDOr sc-gZMcBi ggdJUe"
                >
                  <div
                    class="bp3-input-group sc-gqjmRU kmAUGh"
                  >
                    <input
                      class="bp3-input"
                      name="contests[0].choices[1].numVotes"
                      style="padding-right: 10px;"
                      value=""
                    />
                  </div>
                </div>
              </label>
            </div>
            <p
              class="sc-kGXeez hibFwE"
            >
              Add a new candidate/choice
            </p>
          </div>
        </div>
      </div>
      <div
        class="sc-bdVaJa kcYonD"
      >
        <button
          class="bp3-button sc-bxivhb WkpPV"
          type="button"
        >
          <span
            class="bp3-button-text"
          >
            Add another 
            targeted
             
            contest
          </span>
        </button>
      </div>
    </div>
    <div
      class="sc-bdVaJa kcYonD"
    >
      <button
        class="bp3-button sc-bxivhb WkpPV"
        type="button"
      >
        <span
          class="bp3-button-text"
        >
          Back
        </span>
      </button>
      <button
        class="bp3-button bp3-intent-primary sc-bxivhb WkpPV"
        type="submit"
      >
        <span
          class="bp3-button-text"
        >
          Save & Next
        </span>
      </button>
    </div>
  </form>
</div>
`;
